src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java

```diff
@@ -0,0 +1,172 @@
+        }
+
+        logger.debug("[Stream #{}] Start receiving file #{} from {}, size = {}, ks = '{}', table = '{}'.",
+                     session.planId(), fileSeqNum, session.peer, totalSize, cfs.keyspace.getName(), cfs.getTableName());
+
+        BigTableBlockWriter writer = null;
+
+        try
+        {
+            Set<Component> componentsToWrite = components.stream()
+                                                         .map(p -> Component.parse(p.type.repr))
+                                                         .collect(Collectors3.toImmutableSet());
+
+            writer = createWriter(cfs, totalSize, componentsToWrite);
+            long bytesRead = 0;
+            for (ComponentInfo info : components)
+            {
+                logger.debug("[Stream #{}] About to receive file {} from {} readBytes = {}, componentSize = {}, totalSize = {}",
+                            session.planId(), info.type, session.peer, FBUtilities.prettyPrintMemory(bytesRead),
+                            FBUtilities.prettyPrintMemory(info.length), FBUtilities.prettyPrintMemory(totalSize));
+                writer.writeComponent(info.type, inputPlus, info.length);
+                session.progress(writer.descriptor.filenameFor(Component.parse(info.type.repr)), ProgressInfo.Direction.IN, info.length, info.length);
+                bytesRead += info.length;
+                logger.debug("[Stream #{}] Finished receiving file {} from {} readBytes = {}, componentSize = {}, totalSize = {}",
+                            session.planId(), info.type, session.peer, FBUtilities.prettyPrintMemory(bytesRead),
+                            FBUtilities.prettyPrintMemory(info.length), FBUtilities.prettyPrintMemory(totalSize));
+            }
+
+            return writer;
+        }
+        catch (Throwable e)
+        {
+            logger.error("[Stream {}] Error while reading from stream on ks='{}' and table='{}'.",
+                        session.planId(), cfs.keyspace.getName(), cfs.getTableName(), e);
+            if (writer != null)
+            {
+                writer.abort(e);
+            }
+            throw Throwables.propagate(e);
+        }
+    }
+
+    @SuppressWarnings("resource")
+    protected BigTableBlockWriter createWriter(ColumnFamilyStore cfs, long totalSize, Set<Component> componentsToWrite) throws IOException
+    {
+        Directories.DataDirectory localDir = cfs.getDirectories().getWriteableLocation(totalSize);
+        if (localDir == null)
+            throw new IOException(String.format("Insufficient disk space to store %s", FBUtilities.prettyPrintMemory(totalSize)));
+
+        StreamReceiver streamReceiver = session.getAggregator(tableId);
+
+        assert streamReceiver instanceof CassandraStreamReceiver;
+
+        LifecycleTransaction txn = CassandraStreamReceiver.fromReceiver(session.getAggregator(tableId)).getTransaction();
+
+        // TODO: Is this the correct directory?
```

Comment:
Is this still TODO?

This still doesn't really look right to me. First off we know the size of the sstable, but aren't providing it to Directories, but secondly we know the range of the sstable and presumably it's going to fall into disk boundaries like flushing? How does this interact with partitioning data per disk? Does it violate assumptions made there? @krummas WDYT?

Reply:
Yeah this would violate CASSANDRA-6696 since we write the whole file to the directory with the most space. Things would still work though, compaction will drop the tokens in the correct place. We have the same problem with nodetool refresh/import.

We could have the sending side send the first token of the sstable so that the receiver can drop the received sstable in the correct directory, which would make it correct if the nodes have the same disk count. 

CASSANDRA-10540 will make it work since the sstables would fit in the requested ranges.